- integrate thunknyc profile to cider

jmayaalv opened the pull request:
This is my attempt to solve clojure-emacs/cider#1993 (comment)

This is the companion PR for `cider-nrepl`: https://github.com/clojure-emacs/cider-nrepl/pull/502

Before submitting a PR make sure the following things have been done:

**Replace this placeholder text with a summary of the changes in your PR.
The more detailed you are, the better.**

-----------------

Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines][1]
- [ ] You've added tests (if possible) to cover your change(s)
- [ ] All tests are passing (`make test`)
- [] The new code is not generating bytecode or `M-x checkdoc` warnings
- [ ] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

Thanks!

*If you're just starting out to hack on CIDER you might find this [section of its
manual][2] extremely useful.*

[1]: https://github.com/clojure-emacs/cider/blob/master/.github/CONTRIBUTING.md
[2]: https://cider.readthedocs.io/en/latest/hacking_on_cider/

jmayaalv commented:
This PR is still not ready to be merged. I am having some problems displaying the cider-profile buffer, it's just blank. I am not pretty sure how to debug it. Any help would be appreciated :)

bbatsov commented:
> This PR is still not ready to be merged. I am having some problems displaying the cider-profile buffer, it's just blank. I am not pretty sure how to debug it. Any help would be appreciated :)

Sure thing! I'll take a look at the code and I welcome the other active contributors to take a look as well. :-) /cc @xiongtx @ckoparkar @dpsutton

bbatsov commented on cider-profile.el:
```diff
@@ -0,0 +1,211 @@
+  query)
+
+;;;###autoload
+(define-minor-mode cider-profile-mode
```
I think we don't need this minor mode. Probably we can just put all the commands under `cider-mode` as usual. 
Not to mention those keys are reserved for users, so we can certainly come up with better ones.

bbatsov commented:
Btw, should should by running `M-x checdoc` in this buffer and fixing all the offenses in the docstrings, as those fail the build.

dpsutton commented on cider-profile.el:
```diff
@@ -0,0 +1,211 @@
+
+(defun cider-profile-display-stats (stats-response)
+  (let ((table (nrepl-dict-get stats-response "err")))
+    (if cider-profile-buffer
```
I don't understand why this is a conditional. `cider-profile-buffer` is `defconst`ed to a string.

dpsutton commented on cider-profile.el:
```diff
@@ -0,0 +1,211 @@
+        '()
+        '()
+        '()))
+    (cider-nrepl-send-request
```
this pattern comes up a lot in this code. i'd probably extract it into some type of `cider-message-response-handler` that only takes the lambda of how to case over the successful message. i saw that you handle errors on the nrepl side but i wasn't sure if those would need an error handler on this side.

jmayaalv replied:
Thanks a lot @dpsutton 
this was mostly a copy & paste from https://github.com/thunknyc/nrepl-profile/

I am totally  newby on elisp, so i will do my best :)

dpsutton commented on cider-profile.el:
```diff
@@ -0,0 +1,211 @@
+  (cider-ensure-op-supported "profile-summary")
+  (cider-profile-display-stats
+   (cider-nrepl-send-sync-request (list "op" "profile-summary")))
+  query)
```
i'm not sure what `query` is doing here

jmayaalv replied:
The `query` argument is not using anywhere but the function always accepts it. What's the patter to do it correctly? should i set the `query` argument optional?
This is being done in quiet a few places.

dpsutton replied:
I don't understand what you are trying to accomplish. I don't know of a reason that you need to allow for an argument if you aren't using it. If you are just putting it there to make it interactive, it's fine to remove the argument and just use `(interactive)`. Otherwise, if there's a reason you need some information the function should act on that information.

jmayaalv replied:
Got it! it dint' really understood how `interactive` works.  I think i got it know. Thanks :)

dpsutton replied:
awesome! and thanks for adding this feature!

jmayaalv commented:
I've fixes the problems with the docs :) sorry. Had no idea that i had to run `checkdocs` on the buffer. Learning a lot here :)

I can now see the profile summary which is exciting! Thank you! 
I've also removed the `cider-profile-mode` and added new keybindings.

bbatsov commented on cider-profile.el:
```diff
@@ -0,0 +1,196 @@
+
+;;;###autoload
+(defun cider-profile-message-response-handler (handler &optional buffer)
+  "Default message response handler.  It takes HANDLER to process the message on an optional BUFFER or the current buffer if not specified."
```
I'm surprised this pass the checkdoc test. The second sentence should start on the second line. This applies to all docstrings.

bbatsov commented:
Btw, you might also want to reformat this entire buffer, as right now I see the indent is wrong here and there.

@gonewest818 I've a small idea for a new travis check - whether the files changed have the right indentation or trailing whitespace. Guess we can just compare the committed file to the same file after `indent-region` and some cleanup function (`untabify`, `delete-trailing-whitespace`, etc).

gonewest818 commented:
So, like `cljfmt check` but for elisp.  Yep, that sounds useful.

> On Feb 5, 2018, at 8:44 AM, Bozhidar Batsov <notifications@github.com> wrote:
> 
> Btw, you might also want to reformat this entire buffer, as right now I see the indent is wrong here and there.
> 
> @gonewest818 <https://github.com/gonewest818> I've a small idea for a new travis check - whether the files changed have the right indentation or trailing whitespace. Guess we can just compare the committed file to the same file after indent-region and some cleanup function (untabify, delete-trailing-whitespace, etc).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/clojure-emacs/cider/pull/2197#issuecomment-363143813>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ACWM9roksBFzamUcl-r3mjqoiO0XTXqzks5tRzAKgaJpZM4R4wvy>.
>

xiongtx commented on cider-profile.el:
```diff
@@ -0,0 +1,201 @@
+              "sym" sym)
+        (cider-profile-message-response-handler
+         (lambda (_buffer value)
+           (cond ((equal value "profiled")
```
`cl-case` or `pcase`

bbatsov commented on cider-profile.el:
```diff
@@ -0,0 +1,201 @@
+              "sym" sym)
+        (cider-profile-message-response-handler
+         (lambda (_buffer value)
+           (cond ((equal value "profiled")
```
Here in Emacs you'd normally use `cl-case` or `pcase` to make this code more compact.

xiongtx replied:
Just realized that `cl-case` uses `eql` for comparison, while `pcase` uses `equal`. Strings are _not_ `eql` to each other. So

```emacs-lisp
(cl-case "foo"
  ("foo" "got-foo"))
;; => nil
```

bbatsov commented on cider-profile.el:
```diff
@@ -0,0 +1,201 @@
+           "ns" ns)
+     (cider-profile-message-response-handler
+      (lambda (_buffer value)
+        (cond ((equal value "profiled")
```
Here as well.

bbatsov commented:
@jmayaalv Seems we're almost ready with this. Apart from the tiny feedback about `cl-case`/`pcase`, the only thing missing is a changelog entry. 

Take a look at my feedback about the Clojure part of the PR as well.

bbatsov commented:
I guess we can clean this up after we've merged the initial working version. Thanks for your work on this @jmayaalv! Looking forward to future contributions from you! 😉

bbatsov commented:
I was also reminded that this deserves a page in CIDER's manual - it'd be great if you wrote one when you have some time.

jmayaalv commented:
Thanks a lot for the help guys! 
I will create a new pr this week with the fixes and the entry in the manual! 

Would love to contribute more. If there is something you have in mind that a elisp newby can do i would be glad to help.

stardiviner commented:
@jmayaalv Thanks, I waited this for a long time. Really Thanks a lot.
